pkg/cloudprovider/providers/azure/azure_cache.go

```diff
@@ -19,63 +19,138 @@ package azure
 import (
 	"sync"
 	"time"
-
-	"k8s.io/client-go/tools/cache"
 )
 
-type timedcacheEntry struct {
-	key  string
-	data interface{}
-}
+type getFunc func(key string) (interface{}, error)
+type listFunc func() (map[string]interface{}, error)
+
+// timedCache is a cache with TTL.
+type timedCache struct {
+	store       map[string]interface{}
+	lock        sync.Mutex
+	ttl         time.Duration
+	lastRefresh time.Time
 
-type timedcache struct {
-	store cache.Store
-	lock  sync.Mutex
+	getter getFunc
+	lister listFunc
 }
 
-// ttl time.Duration
-func newTimedcache(ttl time.Duration) timedcache {
-	return timedcache{
-		store: cache.NewTTLStore(cacheKeyFunc, ttl),
+// newTimedcache creates a new timedCache with TTL.
+func newTimedcache(ttl time.Duration, getFunc getFunc, listFunc listFunc) *timedCache {
+	return &timedCache{
+		ttl:    ttl,
+		getter: getFunc,
+		lister: listFunc,
+		store:  make(map[string]interface{}),
 	}
 }
 
-func cacheKeyFunc(obj interface{}) (string, error) {
-	return obj.(*timedcacheEntry).key, nil
+// expired returns true if the cache is expired.
+func (t *timedCache) expired() bool {
+	return t.lastRefresh.Add(t.ttl).Before(time.Now())
 }
 
-func (t *timedcache) GetOrCreate(key string, createFunc func() interface{}) (interface{}, error) {
-	entry, exists, err := t.store.GetByKey(key)
-	if err != nil {
+// Get gets data by key. It returns nil if data not found.
+func (t *timedCache) Get(key string) (interface{}, error) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if t.expired() {
+		if err := t.refresh(""); err != nil {
```

Comment:
Getting one will cause refresh? listing will also cause a refresh? This will cause too many calls to ARM. Can we maintain expiry per key? Listing can refresh per key bases as it sees *expired* per key

Reply:
Also discussed this with @karataliu. Basically, this is a tradeoff between per-key-expiry and list-expiry. 

- per-key-expiry: it supports refresh by per key, but hard to cooperate with LIST, e.g.
  - suppose a large cluster with 200 nodes, and half of the entries are expired
  -  for cache.Get(), it is OK to GET by per key, but the worst case is updating each entry via GET (100 GET calls), which may exceed limitations
  - for cache.List(), LIST is the only safe way to update cache since we don't expect too much GET calls. But updating the entire cache may be time-consuming, e.g. 
     - call LIST first
     - update each cache entry: acquire entry lock and update it if expired.
     - but note that some expired entries may have already been updating inside cache.Get(), so the above lock acquiring should wait a while
- list-expiry: it refresh the cache by LIST, and only refresh single entry by GET when TTL is not expired yet
  - it is safe for both cache.Get() and cache.List(), no flood API calls
  - the drawback is each response body is larger

> nit: GET/LIST means ARM call

Both ways should work for small clusters. But per-key-expiry may still exceed ARM call limitations for large clusters, so I prefer list-expiry way.

@khenidak WDYT?

Reply:
Also note that even with per-key-expiry, we still need a periodical list inside cache. The reason is we don't know whether any entry has been cached or not, but cache.list() should return all possible data

Reply:
ping @khenidak Any comments on this?

Reply:
I still believe per entry expiration is the right way. We don't expect too many LIST calls. and mathematically each LIST call won't invalidate the entire cache (or even the majority). Specially when are updating VMs during attach/detach.